ambry-store/src/main/java/com.github.ambry.store/PersistentIndex.java

```diff
@@ -1010,38 +1041,37 @@ public void write() throws StoreException {
       try {
         if (indexes.size() > 0) {
           // before iterating the map, get the current file end pointer
-          Map.Entry<Long, IndexSegment> lastEntry = indexes.lastEntry();
+          Map.Entry<Offset, IndexSegment> lastEntry = indexes.lastEntry();
           IndexSegment currentInfo = lastEntry.getValue();
-          long currentIndexEndOffsetBeforeFlush = currentInfo.getEndOffset();
-          long logEndOffsetBeforeFlush = logSegment.getEndOffset();
-          if (logEndOffsetBeforeFlush < currentIndexEndOffsetBeforeFlush) {
+          Offset currentIndexEndOffsetBeforeFlush = currentInfo.getEndOffset();
+          Offset logEndOffsetBeforeFlush = log.getEndOffset();
+          if (logEndOffsetBeforeFlush.compareTo(currentIndexEndOffsetBeforeFlush) < 0) {
             throw new StoreException("LogEndOffset " + logEndOffsetBeforeFlush + " before flush cannot be less than "
                 + "currentEndOffSet of index " + currentIndexEndOffsetBeforeFlush, StoreErrorCodes.Illegal_Index_State);
           }
 
           hardDeleter.preLogFlush();
 
           // flush the log to ensure everything till the fileEndPointerBeforeFlush is flushed
-          logSegment.flush();
+          log.flush();
 
           hardDeleter.postLogFlush();
 
-          long lastOffset = lastEntry.getKey();
-          IndexSegment prevInfo = indexes.size() > 1 ? indexes.lowerEntry(lastOffset).getValue() : null;
-          long currentLogEndPointer = logSegment.getEndOffset();
+          IndexSegment prevInfo = indexes.size() > 1 ? indexes.lowerEntry(lastEntry.getKey()).getValue() : null;
+          Offset currentLogEndPointer = log.getEndOffset();
           while (prevInfo != null && !prevInfo.isMapped()) {
-            if (prevInfo.getEndOffset() > currentLogEndPointer) {
+            if (prevInfo.getEndOffset().compareTo(currentLogEndPointer) > 0) {
```

Comment:
Yup, this is the bug. I haven't changed it because it is a big enough change on its own. Have created #501 for this